### PR TITLE
Provide more icons for species and specimen

### PR DESCRIPTION
**Pull request by wow-such-code**
* provides more icon options for species and specimen
* allows users to change the default icons when creating or editing an experiment
* stores selected icon (labels) in the experiment table

image

### CODE REVIEW

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    comboBox.setRenderer(new ComponentRenderer<>(iconResource -> {
+      Span element = new Span();
+      element.addClassName("icon-and-component");
+      AbstractIcon<?> icon = iconResource.getIconResource().createIcon();
+      icon.addClassName("primary");
+      element.add(icon);
+      element.add(iconResource.getLabel());
+      return element;
+    }));
+    return comboBox;
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java:34]:
```suggestion
    comboBox.addValueChangeListener(valueChanged -> valueChanged.getSource()
        .setPrefixComponent(valueChanged.getValue().getIconResource().createIcon()));
    return comboBox;
```
You could also make the combobox change the prefix icon based on the selection 😉

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  /**
+   * Helper class that wraps a Span including an icon around a Combobox with icons. Not used due to
+   * UX concerns, might be useful when more time can be invested into this.
+   */
+  public class ComboWithIcon extends Span {
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java:41]: Please remove. This is not used 😄

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    comboBox.setItemLabelGenerator(
+        BioIcon::getLabel);
+    comboBox.setRenderer(new ComponentRenderer<>(iconResource -> {
+      Span element = new Span();
+      element.addClassName("icon-and-component");
+      AbstractIcon<?> icon = iconResource.getIconResource().createIcon();
+      icon.addClassName("primary");
+      element.add(icon);
+      element.add(iconResource.getLabel());
+      return element;
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java:32]: This is more than 1 line in a lambda. Consider extracting a method `renderIconSelection` or similar here.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java [hunk 4]
```diff
@@ -84,6 +94,24 @@ public String getName() {
...
+
+  public String getAnalyteIconName() {
+    return analyteIconName;
+  }
+
+  public void setIconNames(String speciesIconName, String specimenIconName, String analyteIconName) {
+    this.speciesIconName = speciesIconName;
+    this.specimenIconName = specimenIconName;
+    this.analyteIconName = analyteIconName;
+  }
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java:113]: some sanity checks might be useful here (null, empty, allowed value). Maybe think about providing enums.

File: project-management/src/test/groovy/life/qbic/projectmanagement/application/AddExperimentToProjectServiceSpec.groovy [hunk 5]
```diff
@@ -25,7 +25,7 @@ class AddExperimentToProjectServiceSpec extends Specification {
                 "Pilot",
                 [new OntologyTerm()],
                 [new OntologyTerm()],
-                [new OntologyTerm()])
+                [new OntologyTerm()],"default","default")
```
Comment (KochTobi) [inline project-management/src/test/groovy/life/qbic/projectmanagement/application/AddExperimentToProjectServiceSpec.groovy:28]:
```suggestion
                [new OntologyTerm()],
                "default",
                "default")
```

File: user-interface/frontend/themes/datamanager/components/dialog.css [hunk 6]
```diff
@@ -186,6 +186,22 @@ vaadin-dialog-overlay::part(title) {
...
+  display: inline-flex;
+  justify-content: start;
+  align-items: flex-end;
+  column-gap: var(--lumo-space-m);
+}
+
+.box-flexgrow {
+  flex-grow: 1;
+}
+
```
Comment (KochTobi) [inline user-interface/frontend/themes/datamanager/components/dialog.css:204]: are all of these needed?

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java [hunk 7]
```diff
@@ -38,12 +41,29 @@ public ExperimentalInformationLayout(
         "ontologyTermInformationService must not be null");
     OntologyComboboxFactory ontologyComboboxFactory = new OntologyComboboxFactory(
         ontologyTermInformationService);
+    BioIconComboboxFactory bioIconComboboxFactory = new BioIconComboboxFactory();
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java:44]: probably should be `private final`

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java [hunk 8]
```diff
@@ -61,11 +81,16 @@ public ExperimentalInformationLayout(
         .bind(ExperimentalInformation::getExperimentName,
             ExperimentalInformation::setExperimentName);
 
+    Div speciesRow = new Div(speciesIconBox, speciesBox);
+    speciesRow.addClassName("input-with-icon");
+    Div specimenRow = new Div(specimenIconBox, specimenBox);
+    specimenRow.addClassName("input-with-icon");
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java:87]: As far as I can tell `input-with-icon` is not needed

**Review by KochTobi — APPROVED** [review]
I think it is a good solution, well executed.
Could you provide the links and attribution in javadoc as well so it can be transferred to the about page when ready?

**Review by KochTobi — COMMENTED** [review]
Should these icons also be shown on the project summary? And if so from which experiment are they chosen?

**Review by KochTobi — CHANGES_REQUESTED** [review]
I would like to see some changes applied before approving 😄

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java (hunk 8 above)
Comment (wow-such-code) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/create/ExperimentalInformationLayout.java]: it is, but I renamed it to better reflect what this style is needed for

**Comment by wow-such-code** [comment]
> Should these icons also be shown on the project summary? And if so from which experiment are they chosen?

that is the big question 🤷‍♂️

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java (hunk 3 above)
Comment (wow-such-code) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/create/BioIconComboboxFactory.java]: extracted the icon styling and made the rest significantly shorter

File: user-interface/vite.config.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { UserConfigFn } from 'vite';
+import { overrideVaadinConfig } from './vite.generated';
+
+const customConfig: UserConfigFn = (env) => ({
+  // Here you can add custom Vite parameters
+  // https://vitejs.dev/config/
+});
+
+export default overrideVaadinConfig(customConfig);
```
Comment (KochTobi) [inline user-interface/vite.config.ts:9]: what is this file?

Comment (wow-such-code) [reply user-interface/vite.config.ts]: it is no more 💥

**Review by KochTobi — APPROVED** [review]
LGTM

**Pull request merged**